sklearn/utils/validation.py

```diff
@@ -954,6 +954,16 @@ def check_non_negative(X, whom):
     whom : string
         Who passed X to this function.
     """
-    X = X.data if sp.issparse(X) else X
-    if (X < 0).any():
+    # avoid X.min() on sparse matrix since it also sorts the indices
+    if sp.issparse(X):
+        if X.format in ['lil', 'dok']:
```

Comment:
Should we do this more ducktypingish? `if not hasttr(X, 'data') or not hasattr(X.data, 'min')`?

Reply:
- DOK matrix: no `data` attribute
- LIL matrix: `data` is an array of list, so `X.data.min()` returns the "smallest" list in the sense of lexicographical order.

I don't see a nice way to use ducktyping, do you have something in mind?